### ISSUE TITLE
crash (segfault) in SqliteTableModel::getForeignKeyClause

### ISSUE THREAD

**Issue by NF1198** [sqlitebrowser/sqlitebrowser#995]
### Details for the issue

Sqlitebrowser crashes when viewing an SQL view in the Browse Data tab.

Background: This crash happens when viewing an SQL view in the browser tab.  I can reproduce the crash with a large database but not with a small test DB. To cause the crash, I just open the view, then idle. After a short time (varies), sqlitebrowser will crash.

```
Thread 1 "sqlitebrowser" received signal SIGSEGV, Segmentation fault.
0x00000000005ad826 in QString::size (this=0x8) at /usr/include/qt5/QtCore/qstring.h:231
231         inline int size() const { return d->size; }
```

The crash appears to be occurring in `SqliteTableModel::getForeignKeyClause()`
 (`sqlitetablemodel.cpp:276`)

**Has `obj` or `obj->name` has been freed then accessed?**.

```
#1  0x000000000063a327 in SqliteTableModel::getForeignKeyClause (this=0xd1e6c0, column=2) at /sqlitebrowser/src/sqlitetablemodel.cpp:276
276         if(obj->name().size() && (column >= 0 && column < obj->fields().count()))
```

```
(gdb) print obj
$12 = {value = 0x0, d = 0x0}
```

```
(gdb) print obj->name().size()

Thread 1 "sqlitebrowser" received signal SIGSEGV, Segmentation fault.
0x00000000005ad826 in QString::size (this=0x8) at /usr/include/qt5/QtCore/qstring.h:231
231         inline int size() const { return d->size; }
The program being debugged was signaled while in a function called from GDB.
GDB remains in the frame where the signal was received.
To change this behavior use "set unwindonsignal on".
Evaluation of the expression containing the function
(QString::size() const) will be abandoned.
When the function is done executing, GDB will silently stop.
```

#### I'm opening this issue because:

- [X] DB4S is crashing
- [ ] DB4S has a bug
- [ ] DB4S needs a feature
- [ ] DB4S has another problem

#### I'm using DB4S on:

- [ ] Windows: ( _version:_ ___ )
- [X] Linux: OpenSuse Leap 42.2
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### I'm using DB4S version:

- [X] 3.9.1
- [ ] 3.9.0
- [X] Other: build from source (master: last commit 760ef092cdd26e5cb5eb6222a85612cb2e755540)

#### I have also:

- [ ] Tried out the nightly version: https://github.com/sqlitebrowser/sqlitebrowser#windows
- [X] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
Thanks @NF1198, that's good info.  Foreign key stuff is directly of interest for @MKleusberg, so he'll likely dig into it in the next few days. :smile:

**Comment by justinclift** [comment]
Out of curiosity, is the data set which triggers the crash public?

Just thinking it might be useful for debugging purposes if it is. :wink:

**Comment by NF1198** [comment]
I've attached the dataset. Try browsing `TRIPS_STOPS_EXT`.

It looks like the code causing the crash is:

```
#2  0x000000000063a0d1 in SqliteTableModel::data (this=0xd23420, index=..., role=3) at /sqlitebrowser/src/sqlitetablemodel.cpp:258
258             sqlb::ForeignKeyClause fk = getForeignKeyClause(index.column()-1);
```

commenting out lines 258-261 seems to resolve the crash at the expense of a missing tooltip.

from:
```
        sqlb::ForeignKeyClause fk = getForeignKeyClause(index.column()-1);
        if(fk.isSet())
            return tr("References %1(%2)\nHold Ctrl+Shift and click to jump there").arg(fk.table()).arg(fk.columns().join(","));
        else
            return QString();
```
to
```
        // sqlb::ForeignKeyClause fk = getForeignKeyClause(index.column()-1);
        // if(fk.isSet())
        //     return tr("References %1(%2)\nHold Ctrl+Shift and click to jump there").arg(fk.table()).arg(fk.columns().join(","));
        // else
            return QString();
```

[abq.gtfs.tar.gz](https://github.com/sqlitebrowser/sqlitebrowser/files/931187/abq.gtfs.tar.gz)

**Comment by MKleusberg** [comment]
Thanks for reporting this, @NF1198, and also many thanks for the detailed analysis :+1: That helped a lot :smiley:

I think I have fixed the issue with the latest commit. But just to be sure, can you build the latest version from master and check if it's fixed for you?

**Comment by NF1198** [comment]
Thanks for working on this so quickly @MKleusberg. The crash seems to be fixed with commit d52263fc02b6bdf4033337c4990046ddaa944d3d.

**Comment by justinclift** [comment]
Awesome!  Closing this issue now then.

Thanks for reporting it @NF1198, and giving us the info needed to resolve it promptly. :smile:
